Assets/Shopify/UIToolkit/Components/WebImageCache.cs

```diff
@@ -0,0 +1,156 @@
+ namespace Shopify.UIToolkit {
+    using UnityEngine;
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// A web resource that is cached against a Last-Modified timestamp.
+    /// </summary>
+    public struct CachedWebResource<T> {
+        public readonly string LastModifiedTimestamp;
+        public readonly T Data;
+
+        public CachedWebResource(string lastModifiedTimestamp, T data) {
+            LastModifiedTimestamp = lastModifiedTimestamp;
+            Data = data;
+        }
+    }
+
+    /// <summary>
+    /// This class implements a LRU cache for images (PNG/JPEG) that were downloaded from the web.
+    /// </summary>
+    public class WebImageCache {
+        /// <summary>
+        /// The default estimated memory size limit in bytes.
+        /// </summary>
+        public const int DEFAULT_MEMORY_SIZE_LIMIT = 20971520; // 20 MB in bytes
+
+        private Dictionary<string, CachedWebResource<Texture2D>> _urlTextureCache = new Dictionary<string, CachedWebResource<Texture2D>>();
+        private LinkedList<string> _recentlyUsed = new LinkedList<string>();
+
+        private static WebImageCache _sharedCache;
+
+        /// <summary>
+        /// Shared WebImageCache instanced.
+        /// </summary>
+        /// <returns>A singleton instance of the WebImageCache.</returns>
+        public static WebImageCache SharedCache {
+            get {
+                _sharedCache = _sharedCache ?? new WebImageCache(DEFAULT_MEMORY_SIZE_LIMIT);
+                return _sharedCache;
+            }
+        }
+
+        /// <summary>
+        /// The estimated size limit in bytes of how much memory we want to use for the cache (readonly).
+        /// </summary>
+        /// <returns></returns>
+        public int MemorySizeLimit { get; private set; }
+
+        /// <summary>
+        /// The current estimated size the cache takes up in memory (readonly)
+        /// </summary>
+        /// <returns></returns>
+        public int EstimatedMemorySize { get; private set; }
+
+        public WebImageCache(int memorySizeLimit) {
```

Comment:
if there is a singleton, should the constructor be public?

Reply:
🤔

Reply:
as in shouldnt the singleton method be the only access point, therefore the constructor be private lol... bad rhetoric

Reply:
Heh yea it should be private 😶